soroban-react-dapp/contracts/greeting/src/lib.rs

```diff
@@ -1,24 +1,96 @@
 #![no_std]
-use soroban_sdk::{contract, contractimpl, Env, Symbol, symbol_short, String};
-
-const TITLE: Symbol = symbol_short!("TITLE");
-
+use soroban_sdk::{contract, contracterror, contractimpl, contracttype, Address, Env, String, Vec};
 
 #[contract]
 pub struct TitleContract;
 
+#[contracterror]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum Error {
+    Unauthorized = 1,
+    AlreadyInitialized = 2,
+}
+
 #[contractimpl]
 impl TitleContract {
+    // initialize the contract and set the admin
+    pub fn init(env: Env, admin: Address) -> Result<(), Error> {
+        let storage = env.storage().instance();
+        admin.require_auth();
+        if storage.has(&Assets::Admin) {
+            return Err(Error::AlreadyInitialized);
+        }
+        storage.set(&Assets::Admin, &admin);
+        Ok(())
+    }
 
-    pub fn set_title(env: Env, title: String) {
-                env.storage().instance().set(&TITLE, &title)
+    // set the title only available editors
+    pub fn set_title(env: Env, user: Address, title: String) -> Result<(), Error> {
+        user.require_auth();
+        let storage = env.storage().instance();
+        let admin: Address = storage.get(&Assets::Admin).unwrap();
+        let editors: Vec<Address> = storage.get(&Assets::Editors).unwrap_or(Vec::new(&env));
+        if editors.contains(&user) || user.eq(&admin) {
+            env.storage().instance().set(&Assets::Title, &title);
+            Ok(())
+        } else {
+            Err(Error::Unauthorized)
+        }
     }
 
+    // read the title
     pub fn read_title(env: Env) -> String {
-        env.storage().instance().get(&TITLE)
+        env.storage()
+            .instance()
+            .get(&Assets::Title)
             .unwrap_or(String::from_str(&env, "Default Title"))
     }
-    
-} 
+
+    /// ***** Address Management ***** ///
+
+    // add wallet address for editors
+    pub fn add_editor(env: Env, new_editor: Address) {
```

Comment:
This function can be attacked with a DOS. But its ok because it can only be called by the admin

Reply:
okay.